Desktop: Add external editor actions to the note context menu.

mic704b opened the pull request:
~Also start up external editor on note double click.~

These changes enhance user experience by placing the actions where they feel natural.

mic704b commented:
The proposed changes are to add external editor actions to the note context menu.  In my experience, this is a place where the user might naturally expect to find them.

When right-click on a note, "Edit in external editor" is a new option.  If the note is in the watching state already, then the new option is "Stop watching external editor".  If two or more notes are selected, then the option is unavailable.

In addition, double-click on the note in the note list launches the external editor.  If the note is in the watching state, then double click does nothing.

I could not find tests relevant to this part of the application. Please advise if there are test cases to be updated.

tessus commented on ElectronClient/app/gui/NoteList.jsx:
```diff
@@ -117,6 +118,15 @@ class NoteListComponent extends React.Component {
+				});
+			}
+		};
+
```
I'm not sure, if this is a good idea. We might want to use a double-click for a different action in the future. But this is Laurent's decision.

mic704b replied:
Hi Tessus, thanks for your comments.

Apologies, I didn't bring up the idea in the forum before,  I read in the contribution instructions that small changes could be proposed on risk.

The double-click action mimics the behaviour of another well known note-taking tool I have used. It seemed natural to me, but perhaps that's just because I am familiar with it.

Actually, many times in Joplin I double clicked on notes (to open, and of course nothing happened), that's what prompted me to try to add it!

In any case, that part can be easily reverted if necessary

tessus replied:
No worries, you are right. Smaller PRs can be submitted without discussing them on the forum first. It's just less or no risk when doing so.

The double-click feedback was just a personal opinion, so we have to wait for @laurent22's input anyway. Maybe there's a bettter action for double-click, but maybe not.

It would be great to have some sort of action preferences to define such action shortcuts, but this might be more complex thus an idea for the future.

mic704b replied:
Good idea

laurent22 commented on ElectronClient/app/gui/utils/NoteListUtils.js:
```diff
@@ -49,6 +49,34 @@ class NoteListUtils {
+			} else {
+				menu.append(
+					new MenuItem({
+						label: _('Stop watching external editor'),
```
Maybe just "Stop external editing"?

mic704b replied:
Yes, will update.

laurent22 commented:
Looks good but I prefer we reserve the use of double-click for now, as not everyone has an external editor configured. I'd expect double-clicking opens the note in a new window, not necessarily in an external editor.

mic704b commented:
Agreed. Will remove the double-click.

mic704b commented:
While testing this, I found a pre-existing bug in the context menu; when you right-click on a note in the list, the context menu is shown for the note under the mouse, but the action is applied to the currently selected note.

Perhaps right-clicking a note in the list should also select it if it isn't already selected? This probably needs discussion, should I create a separate issue for that?

mic704b commented:
There was a bug (as described above) and I have fixed it for this feature in the last commit.  To fix it, I moved the launching of the external editor to the main screen object, because the current note may not be the one that is being acted upon.

The same bug exists for the pdf export and maybe other actions in the context menu, I haven't checked them all. I propose to create an issue to fix those separately.

tessus commented:
> To fix it, I moved the launching of the external editor to the main screen object, because the current note may not be the one that is being acted upon.

I am not sure if this was necessary. e.g. `Delete` works and is not in the main screen object either.
I think this should be fixed differently.

But `export to pdf` has the issue you described. This certainly should be fixed.

mic704b commented:
> I am not sure if this was necessary. e.g. `Delete` works and is not in the main screen object either.
> I think this should be fixed differently.

Yes, that is correct.

I can implement the fix in `NoteListUtils`. Would that be acceptable? It seems sensible from a design point of view. I'm still not 100% across the architecture so am grateful for any guidance.

mic704b commented:
For reference, I used the `Add remove tags`  as the model for the fix, which is implemented in the `MainScreen`.

mic704b commented:
I have reverted the changes made in `MainScreen` and re-implemented the fix back in the `NoteList`.

mic704b commented:
Please see #2255 prior to this. If it is acceptable, then I would like to update this before review, to use the same method.

laurent22 commented:
Looks good, thanks @mic704b!

mic704b commented:
Thank you!
